hicommonwealth/commonwealth
# Migrated threads state to react query

## Link to Issue
Closes: #4588, #4541
Possibly fixed: https://github.com/hicommonwealth/commonwealth/issues/4606

## Description of Changes

### Removed these API calls from frontend
- `/threadsUsersCountAndAvatars`- which was making API calls unnecessarily, and we weren't even using that API response. @rbennettcw / @jnaviask  can we remove this completely from API, I think this would be better to avoid any possible tech debt in future?
- `/reactionsCounts` - which returned reaction counts for threads, proposals and comments, we already store reactions for each of those in their respective objects -- so calling this API per thread/proposal/comment was just an unnecessary load on the app -- removed this unnecessary logic and used the reaction counts already available in thread and proposal object. For comment we still need to optimize -- I think we can do that in a followup ticket as it related to comments and not threads directly. @rbennettcw / @jnaviask  I suggest we also remove this, as this is totally redundant

### Thread models and counters in frontend
- Consolidated the thread controller `modelFromServer` and thread interface (`new Thread()`), which makes the whole thread object more predictable when used in the frontend
- The `/bulkOffchain` API returns the total threads count and total voting threads count when a chain is init. This count is not being stored in zustand in the `useEXCEPTION_CASE_threadCountersStore`, as the name suggests we should not store API responses in zustand store but this is an exception case -- @masvelio for thoughts on this and any ways to improve this

### Added caches for these API calls and responses
Added separate caches for all FETCH thread APIs and all thread API configurations, these are all the thread apis in cache now

1. `/threads?bulk=true` -- this can be configured with query params like `/threads?bulk=true&topicId=1234&toDate={date}&page={page}&stage={stage}&includePinnedThreads={includePinnedThreads}&fromDate={fromDate}&limit={limit}&chain={chain}&orderBy={orderBy}`. All of these query params can be used independently or with any configuration i.e there are 9 query params all of which can be configured. Any response from any configuration will be stored in the cache separately and updates will be propagated to all API response configurations in the cache --- This is very helpful as when we switch b/w communities data won't be lost which would speed up over community switching and the overall thread experience
2. `/threads?active=true`-- this can be configured with query params. Any response from any configuration will be stored in the cache separately and updates will be propagated to all API response configurations in the cache. I added the same things for this api/response/cache as I mentioned in the point above.
3. `/threads?thread_ids=[]` -- this endpoint and response is different from the above apis, even tho the verb is same. I added the same things for this api/response/cache as I mentioned in the point above.



### Moved all thread CRUD API calls from the old thread controller to the new react query implementation

All of the existing thread APIs were moved, exceptions are added in the last

**APIs for fetching threads**

- `GET /threads?bulk=true` - the response from this API is used in the thread listing pages for each topic and it return a paginated response and is different from `GET /threads?active=true` and  `GET /threads?thread_ids=[]` APIs
- `GET /threads?active=true` - the response from this API is only used in the `/overview` page --  this is not a different query param but a completely different API from `GET /threads?bulk=true` and  `GET /threads?thread_ids=[]` -- cc: @rbennettcw maybe we can consolidate this?
- `GET /threads?thread_ids[]` -  used to get a thread by Ids - there can be many ids -- this is not a different query param but a completely different API from `GET /threads?bulk=true` and  `GET /threads?active=true` -- the response from this api is only used in the individual thread page aka thread details page.
- `GET /linking/getLinks` - used to get thread by snapshot link - and this also gets stored in the cache

**APIs for updating thread fields**

- `POST /linking/addThreadLinks` - for adding a link thread/discussion to an existing thread 
- `DELETE /linking/deleteLinks` - for removing a link thread/discussion to an existing thread
- `POST /updateThreadPrivacy` --  used to lock/unlock a thread.
- `POST /updateThreadStage` -- used to update thread stage.
- `POST /updateTopic` -- used to update thread topic.
- `POST /updateThreadPinned` -- used to update thread pinned status.
- `POST /threads/${threadId}/spam` -- used to update thread spam status.
- `POST /addEditors` -- used to add collaborators to thread
- `POST /deleteEditors` -- used to removed collaborators from thread
- `PATCH /threads/${threadId}` -- used to update a thread title/description and basic info

When a thread field changes we propagated the updated state to all cache and since RQ is reactive we don't have to worry about "rendering" ourselves. 

### Other things we updated that effect thread cache 
- Added logic to update thread reactions in the thread object itself, instead of using the `/reactionCounts` API per thread. This also improves performance. Also added logic in thread reaction methods to update cache everywhere when reactions are changed.
- `POST /threads/${threadId}/comments` -- adds a comment, when we do that we also update thread total comments counts
- `POST /comments/${commentId}` - deletes a comment, , when we do that we also update thread total comments counts.


### Thread APIs that seems broken
- `PUT/DELETE /threads/${threadId}/spam` - this API is not working for some reason, in discussions with Ryan
- `POST /threads/${threadId}/archive and /threads/${threadId}/unarchive` - Discussed with @rbennettcw and it looks like @Miaplacidus is working on it in https://github.com/hicommonwealth/commonwealth/pull/4338

### Thread APIs that can be moved to react query
~~- `GET /threads?search=search text` -- @masvelio I was undecided if we should move this to react query, what do you suggest.~~ This was already done


## Test Plan
All of these should work
- updating a thread title and description
- adding/removing thread collaborators/editors
- updating thread stage
- locking/unlocking a thread
- pinning/unpinning a thread
- changing a thread topic
- add/delete thread links
- adding/deleting polls
- adding/deleting proposals
- added/removing comments updates numberOfComments for thread (the little comment button with the comments count)
- add/delete reaction to a thread

## Deployment Plan
N/A

## Other Considerations
N/A
